Brief description for cloud-config-sync controller in dev docs (#72)
openshift/cluster-cloud-controller-manager-operator: docs/dev/cloud-config-sync.md

# Managing cloud-config for CCMs

## Intro

Some cloud providers, such as [Azure](https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs/) or [vSphere](https://cloud-provider-vsphere.sigs.k8s.io/cloud_config.html), require a config file, which contains various platform-specific parameters (e.g. api endpoints, resource group name, and so on).

> In OpenShift this config is stored in a config map defined during the installation procedure and managed by the [cluster-config-operator](https://github.com/openshift/cluster-config-operator).

Reviewer: minor nit
```suggestion
In OpenShift this config is stored in a ConfigMap defined during the installation procedure and managed by the [cluster-config-operator](https://github.com/openshift/cluster-config-operator).
```

Author: fixed

> There are two places where this config map is stored on a running cluster at the moment (OCP 4.9):

Reviewer: minor nit
```suggestion
There are two places where this ConfigMap is stored on a running cluster at the moment (OCP 4.9):
```

# Managing cloud-config for CCMs

# Intro

Some cloud providers,
such as [Azure](https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs/) or [vSphere](https://cloud-provider-vsphere.sigs.k8s.io/cloud_config.html),

> require config file which contains various platform-specific parameters (e.g. api endpoints, resource group name or so).

Reviewer: ```suggestion
require a config file which contains various platform-specific parameters (e.g. api endpoints, resource group name or so).
```

> In OpenShift terms this config represented as ConfigMap defined during the installation procedure and managing by [cluster-config-operator](https://github.com/openshift/cluster-config-operator).

Reviewer: ```suggestion
In OpenShift terms this config is represented as ConfigMap defined during the installation procedure and managing by the [cluster-config-operator](https://github.com/openshift/cluster-config-operator).
```

> There are two places where this config map stored on a running cluster at the moment (OCP 4.8):

Reviewer: ```suggestion
There are two places where this config map is stored on a running cluster at the moment (OCP 4.8):
```

1. `kube-cloud-config` ConfigMap in `openshift-config-managed` namespace

> 2. ConfigMap with arbitrary name in `openshift-config` namespace. Such name might be taken from `cluster` Infrastructure resource spec.

Reviewer: is this an automatically generated name?

Author: Don't know to be honest. 
https://docs.openshift.com/container-platform/4.7/rest_api/config_apis/infrastructure-config-openshift-io-v1.html#specification
Didn't find from where it comes. Judging by Infrastructure resource schema - it might be arbitrary string.

Reviewer 2: It is a field in the spec of the user managed resource. It is definitely something that could be different from installation to installation.

> This ConfigMap should be copied from one of the places described above and keep in sync within CCCMO managed namespace for further mount onto cloud provider pods.

Reviewer: ```suggestion
This ConfigMap should be copied from one of the places described above and kept in sync within CCCMO managed namespace for further mounting onto cloud provider pods.
```

> For such purposes separate controller within CCCMO was introduced (see `controllers/cloud_config_sync_controller.go`).

Reviewer: ```suggestion
For such purposes a separate controller within CCCMO was introduced (see `controllers/cloud_config_sync_controller.go`).
```

> # Managing cloud-config for CCMs
> 
> # Intro
> 
> Some cloud providers,
> such as [Azure](https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs/) or [vSphere](https://cloud-provider-vsphere.sigs.k8s.io/cloud_config.html),
> require a config file which contains various platform-specific parameters (e.g. api endpoints, resource group name or so).
> 
> In OpenShift terms this config is represented as ConfigMap defined during the installation procedure and managing by the [cluster-config-operator](https://github.com/openshift/cluster-config-operator).
> 
> There are two places where this config map is stored on a running cluster at the moment (OCP 4.8):
> 1. `kube-cloud-config` ConfigMap in `openshift-config-managed` namespace
> 2. ConfigMap with arbitrary name in `openshift-config` namespace. Such name might be taken from `cluster` Infrastructure resource spec.
> 
> 
> This ConfigMap should be copied from one of the places described above and kept in sync within CCCMO managed namespace for further mounting onto cloud provider pods.
> For such purposes a separate controller within CCCMO was introduced (see `controllers/cloud_config_sync_controller.go`).
> 
> # Implementation Description
> Implementation is being inspired by `library-go`.
> 
> Brief description:
> 
>     1. In case of changing/deletion/creation one of the following resources
>         - `kube-cloud-config` ConfigMap in `openshift-config-managed` namespace
>         - `cloud-config` ConfigMap in CCCMO managed namespace
>         - `cluster` Infrastructure resource
>        
>     2. Sync CCCMO's `cloud-config` content with `openshift-config-managed/kube-cloud-config`
>     
>     3. If `openshift-config-managed/kube-cloud-config` does not exists - fallback to sync with ConfigMap from `openshift-config` namespace
>         - during sync procedure replace key in target ConfigMap to `cloud.conf` which is default one for OpenShift
> 
> # Links
> - [library-go implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L82)
> - [cluster-config-operator repository](https://github.com/openshift/cluster-config-operator)

Reviewer 3: ```suggestion
# Managing cloud-config for CCMs

## Intro

Some cloud providers, such as [Azure](https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs/) or [vSphere](https://cloud-provider-vsphere.sigs.k8s.io/cloud_config.html), require a config file, which contains various platform-specific parameters (e.g. api endpoints, resource group name, and so on).

In OpenShift this config is stored in a config map defined during the installation procedure and managed by the [cluster-config-operator](https://github.com/openshift/cluster-config-operator).

There are two places where this config map is stored on a running cluster at the moment (OCP 4.9):
1. `kube-cloud-config` ConfigMap in `openshift-config-managed` namespace.
2. ConfigMap with an arbitrary name in `openshift-config` namespace. Such name might be taken from the `cluster` Infrastructure resource spec.

This ConfigMap should be copied from one of the places described above and kept in sync within the CCCMO managed namespace for further mounting onto cloud provider pods. For such purposes a separate controller within CCCMO was introduced (see `controllers/cloud_config_sync_controller.go`).

## Implementation Description

Implementation is being inspired by [library-go](https://github.com/openshift/library-go).

The controller performs a sync of the CCM's `cloud-config` content with `openshift-config-managed/kube-cloud-config` in case of changing/deletion/creation one of the following resources:
        - `kube-cloud-config` ConfigMap in `openshift-config-managed` namespace;
        - `cloud-config` ConfigMap in the CCCMO managed namespace;
        - `cluster` Infrastructure resource.

If `openshift-config-managed/kube-cloud-config` does not exists - the controller fallbacks to sync with the ConfigMap from `openshift-config` namespace. Also during the sync procedure it replaces key in the target ConfigMap to `cloud.conf`, which is default one for OpenShift.

## Links
- [library-go implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go#L82)
- [cluster-config-operator repository](https://github.com/openshift/cluster-config-operator)
```
